src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_DecisionDagRewriter.cs

```diff
@@ -410,53 +412,511 @@ private void GenerateTest(BoundExpression test, BoundDecisionDagNode whenTrue, B
+                                var C = y.Right;
+                                var D = right;
+                                return y.WithLeftAndRight(x, new RelationalDispatch(syntax, value, op, C, D));
+                            }
+                        }
+                        else if (right.Height == left.Height + 2)
+                        {
+                            var rightDispatch = (RelationalDispatch)right;
+                            if (rightDispatch.Right.Height == left.Height)
+                            {
+                                //
+                                //     x
+                                //    / \                   y
+                                //   A   z                 / \
+                                //      / \      -->      x   z
+                                //     y   D             /|   |\
+                                //    / \               A B   C D
+                                //   B   C
+                                //
+                                var A = left;
+                                var z = rightDispatch;
+                                var y = (RelationalDispatch)z.Left;
+                                var B = y.Left;
+                                var C = y.Right;
+                                var D = z.Right;
+                                return y.WithLeftAndRight(new RelationalDispatch(syntax, value, op, A, B), z.WithLeftAndRight(C, D));
+                            }
+                            else
+                            {
+                                Debug.Assert(rightDispatch.Left.Height == left.Height);
+                                //
+                                //     x
+                                //    / \                   y
+                                //   A   y                 / \
+                                //      / \      -->      x   z
+                                //     B   z             /|   |\
+                                //        / \           A B   C D
+                                //       C   D
+                                //
+                                var A = left;
+                                var y = rightDispatch;
+                                var B = y.Left;
+                                var z = y.Right;
+                                return y.WithLeftAndRight(new RelationalDispatch(syntax, value, op, A, B), z);
+                            }
+                        }
+                        #endregion Rebalance the top of the tree if necessary
+
+                        // that should have been sufficient to balance the tree.
+                        Debug.Assert(Math.Abs(left.Height - right.Height) < 2);
+                        return new RelationalDispatch(syntax, value, op, left: left, right: right);
+                    }
+                }
+            }
+
+            private void LowerValueDispatchNode(
```

Comment:
nit: consider inlining `LowerValueDispatchNode` since it's simple, and its single caller is also simple. Then we can remove `Core` suffix from next method